### ISSUE TITLE
Unify matrix libraries

### ISSUE THREAD

**Issue by ctrueden** [fiji/fiji#135]
Several projects throughout the Fiji stack make use of matrix libraries. Often Jama is used. 

@kephale suggested that we could use a more modern library like [ojAlgo](http://ojalgo.org/) or [jblas](http://jblas.org/) or [Vectorz](https://github.com/mikera/vectorz) or [ejml](http://ejml.org/). Or even Clojure's [core.matrix](https://github.com/mikera/core.matrix) as a Java library, which allows switching between backends, though that might have some logistical difficulties.

In any case, would be nice to unify the Fiji stack beneath a single approach to matrices. Likely this unification would actually be centered under ImgLib2, and then the Fiji stack would consume that.

**Comment by tpietzsch** [comment]
Good idea, I support that! I had a look at ejml and I liked it. Pure java and no dependencies. The same seems to be the case for ojAlgo, but I haven't looked at it in detail. I wouldn't use jblas because it wraps native code and I think the overhead of a JNI call for e.g. a 3D matrix vector multiplication is not worth it. No idea about Vectorz, has some dependencies but don't know how bad it is. Using a Clojure lib sounds like we would pull in a lot of dependencies. I don't like that. I would go with ejml, ojAlgo, or possibly Vectorz.

**Comment by kephale** [comment]
Yes, that is mostly fair about core.matrix. FWIW there is only 1 dependency in core.matrix: Clojure (which is already in FIJI, although certainly not in imglib2 stuff, etc.). Then each implementation may have its own dependencies, but that is expected. If there is some debate about which are the most expensive calculations, then it might be worth taking this a little more seriously. Otherwise, it doesn't make sense to bother with.

Vectorz was built from EJML.

I've used jblas and ejml for some simulations that ran continuously on a few dozen machines continuously for months, and was happier with EJML. 

I used ojAlgo for a while after that because it did better on speed tests, but didn't do any long term tests with it. 

When it comes to pure java, ojAlgo generally seems to be winning in performance: http://lessthanoptimal.github.io/Java-Matrix-Benchmark/
Latest report is: 
http://lessthanoptimal.github.io/Java-Matrix-Benchmark/runtime/2015_07_XeonQuad/
However, when it comes to decomposition and solving the results aren't so clear. Since I haven't done deep profiling of the current JAMA/etc. usage, perhaps someone can weigh in on what the most heavily used matrix operations are.

**Comment by kephale** [comment]
Sounds like ojAlgo is the favored option!

**Comment by StephanPreibisch** [comment]
ojAlgo seems to do well according to the tests, and at the same time has the smallest memory footprint: http://lessthanoptimal.github.io/Java-Matrix-Benchmark/memory/2013_10/

Both are interface-based so it should actually be possible to use/wrap ImgLib2 datastructures as matrices. So I would support ojAlgo.

**Comment by tpietzsch** [comment]
+1

**Comment by bogovicj** [comment]
Late to the party, but very happy this conversation is happening - of the libraries mentioned, I've used ejml and have been quite happy with it.  

But after seeing the performance of ojAlgo linked to by @kephale and @StephanPreibisch, I'm now interested in giving it a try.  If and when I do, I'll weigh back in.

**Comment by stelfrich** [comment]
For the record.

@dietzc in Gitter:
> [nd4j](http://nd4j.org/)? dont know

**Comment by stelfrich** [comment]
I thought I had already added a comment about that, well, seems like I did not :frowning:

When looking into https://github.com/imglib/imglib2/issues/137, I also did a short survey on what functionality is used in the fiji organization:
- `Matrix.transpose()` = matrix transpose
- `Matrix.solve()` = solve Ax = B
- `Matrix.eig()` = eigenvalue decomposition
- `Matrix.normF()` = Frobenius norms (also other norms)

vecmath seems to be mostly used for data handling (`Point3x`, `Matrix4x`, `Vector3x`, etc.) in combination with `java3d.Transform3D`.
